tests/data_sources/test_uspto.py

```diff
@@ -0,0 +1,487 @@
+import unittest
+import sqlite3
+
+from ..test_dir import add_src_dir, td
+
+add_src_dir()
+
+from ..common import PopulateQueries, record_count
+from alexandria3k.common import ensure_unlinked
+from alexandria3k.data_sources import uspto
+from alexandria3k.file_xml_cache import FileCache
+from alexandria3k import debug
+
+
+DATABASE_PATH = td("tmp/uspto.db")
+ATTACHED_DATABASE_PATH = td("tmp/attached_uspto.db")
+
+
+def populate_attached():
+    """Create and populate an attached database"""
+    ensure_unlinked(ATTACHED_DATABASE_PATH)
+    attached = sqlite3.connect(ATTACHED_DATABASE_PATH)
+    attached.execute("CREATE TABLE s_us_patents(type)")
+    attached.execute("INSERT INTO s_us_patents VALUES('design')")
+    attached.commit()
+    attached.close()
+
+
+class TestUsptoPopulateVanilla(PopulateQueries):
+    @classmethod
+    def setUpClass(cls):
+        if os.path.exists(DATABASE_PATH):
+            os.unlink(DATABASE_PATH)
+
+        FileCache.parse_counter = 0
+        cls.uspto = uspto.Uspto(
+            td("data/April 2023 Patent Grant Bibliographic Data")
+        )
+        cls.uspto.populate(DATABASE_PATH)
+        cls.con = sqlite3.connect(DATABASE_PATH)
+        cls.cursor = cls.con.cursor()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.con.close()
+        os.unlink(DATABASE_PATH)
+
+    def test_import(
+        self,
+    ):
+        result = TestUsptoPopulateVanilla.cursor.execute(
+            f"SELECT Count(*) from us_patents"
+        )
+        (count,) = result.fetchone()
+        self.assertEqual(count, 10)
+
+    def test_type(
+        self,
+    ):
+        result = TestUsptoPopulateVanilla.cursor.execute(
+            f"""SELECT type FROM us_patents
+            WHERE filename='USPP034694-20221025.XML'"""
+        )
+        (type,) = result.fetchone()
+        self.assertEqual(type, "plant")
+
+    def test_primary_examiner(
+        self,
+    ):
+        result = TestUsptoPopulateVanilla.cursor.execute(
+            f"""SELECT primary_examiner_firstname, primary_examiner_lastname FROM us_patents
+            WHERE filename='USRE049258-20221025.XML'"""
+        )
+        (name, lastname) = result.fetchone()
+        self.assertEqual(name, "Catherine M")
+        self.assertEqual(lastname, "Tarae")
+
+    def test_counts(self):
+        self.assertEqual(self.record_count("us_patents"), 10)
+        self.assertEqual(self.record_count("icpr_classifications"), 24)
+
+        self.assertEqual(
+            self.record_count(
+                """(SELECT DISTINCT type
+          FROM us_patents)"""
+            ),
+            4,
+        )
+
+        self.assertEqual(
+            self.record_count(
+                """(SELECT DISTINCT primary_examiner_firstname
+          FROM us_patents)"""
+            ),
+            8,
+        )
+
+        self.assertEqual(
+            self.cond_count("us_patents", "series_code = '17'"), 3
+        )
+        self.assertEqual(FileCache.parse_counter, 10)
```

Comment:
Please also add a FileCache test that will show that only a singly file has been read (e.g. with SQL `LIMIT 1`.

Reply:
Added both testing XML parsing and Zip file decompressing on a LIMIT 1 case.